File: app/src/main/java/com/elementary/tasks/reminder/filters/ReminderFilterController.java
package com.elementary.tasks.reminder.filters;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.elementary.tasks.reminder.models.Reminder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Copyright 2017 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class ReminderFilterController {

    @NonNull
    private FilterValue<String> searchValue = new FilterValue<>();
    @NonNull
    private FilterValue<List<String>> groupValue = new FilterValue<>();
    @NonNull
    private FilterValue<Integer> typeValue = new FilterValue<>();
    @NonNull
    private FilterValue<Integer> statusValue = new FilterValue<>();
    @NonNull
    private FilterValue<DateFilter.DateRange> rangeValue = new FilterValue<>();

    @NonNull
    private List<Reminder> original = new ArrayList<>();
    @Nullable
    private FilterCallback<Reminder> mCallback;
    @Nullable
    private ObjectFilter<Reminder> mFilter;

    public ReminderFilterController(@Nullable FilterCallback<Reminder> callback) {
        this.mCallback = callback;
        initFilters();
    }

    private void initFilters() {
        SearchFilter searchFilter = new SearchFilter(null);
        searchValue.subscribe(searchFilter);

        DateFilter dateFilter = new DateFilter(searchFilter);
        rangeValue.subscribe(dateFilter);

        GroupFilter groupFilter = new GroupFilter(dateFilter);
        groupValue.subscribe(groupFilter);

        TypeFilter typeFilter = new TypeFilter(groupFilter);
        typeValue.subscribe(typeFilter);

        StatusFilter statusFilter = new StatusFilter(typeFilter);
        statusValue.subscribe(statusFilter);

        this.mFilter = statusFilter;
    }

    public void setSearchValue(@Nullable String value) {
        if (value == null) {
            searchValue.setValue("");
        } else {
            searchValue.setValue(value);
        }
        onChanged();
    }

    public void setGroupValue(@Nullable String value) {
        if (value == null) {
            groupValue.setValue(Collections.singletonList(""));
        } else {
            groupValue.setValue(Collections.singletonList(value));
        }
        onChanged();
    }

    public void setGroupValues(@NonNull List<String> value) {
        groupValue.setValue(value);
        onChanged();
    }

    public void setTypeValue(int value) {
        typeValue.setValue(value);
        onChanged();
    }

    public void setStatusValue(int value) {
        statusValue.setValue(value);
        onChanged();
    }

    public void setRangeValue(int value) {
        switch (value) {
            case 0:
                this.rangeValue.setValue(DateFilter.DateRange.ALL);
                break;
            case 1:
                this.rangeValue.setValue(DateFilter.DateRange.PERMANENT);
                break;
            case 2:
                this.rangeValue.setValue(DateFilter.DateRange.TODAY);
                break;
            case 3:
                this.rangeValue.setValue(DateFilter.DateRange.TOMORROW);
                break;
        }
        onChanged();
    }

    @NonNull
    public List<Reminder> getOriginal() {
        return original;
    }

    public void setOriginal(@NonNull List<Reminder> original) {
        this.original = original;
        onChanged();
    }

    private void onChanged() {
        List<Reminder> list = new ArrayList<>();
        for (Reminder reminder : original) {
            if (mFilter != null) {
                if (mFilter.filter(reminder)) list.add(reminder);
            } else list.add(reminder);
        }
        if (mCallback != null) mCallback.onChanged(list);
    }

    public void remove(@NonNull Reminder reminder) {
        if (original.contains(reminder)) {
            original.remove(reminder);
            onChanged();
        }
    }
}

File: app/src/main/java/com/elementary/tasks/reminder/lists/filters/ReminderFilterController.kt
package com.elementary.tasks.reminder.lists.filters

import com.elementary.tasks.core.data.models.Reminder

import java.util.ArrayList
import java.util.Collections

/**
 * Copyright 2017 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class ReminderFilterController(private val mCallback: FilterCallback<Reminder>?) {

    private val searchValue = FilterValue<String>()
    private val groupValue = FilterValue<List<String>>()
    private val typeValue = FilterValue<Int>()
    private val statusValue = FilterValue<Int>()
    private val rangeValue = FilterValue<DateFilter.DateRange>()

    private var original: MutableList<Reminder> = ArrayList()
    private var mFilter: ObjectFilter<Reminder>? = null

    init {
        initFilters()
    }

    private fun initFilters() {
        val searchFilter = SearchFilter(null)
        searchValue.subscribe(searchFilter)

        val dateFilter = DateFilter(searchFilter)
        rangeValue.subscribe(dateFilter)

        val groupFilter = GroupFilter(dateFilter)
        groupValue.subscribe(groupFilter)

        val typeFilter = TypeFilter(groupFilter)
        typeValue.subscribe(typeFilter)

        val statusFilter = StatusFilter(typeFilter)
        statusValue.subscribe(statusFilter)

        this.mFilter = statusFilter
    }

    fun setSearchValue(value: String?) {
        if (value == null) {
            searchValue.setValue("")
        } else {
            searchValue.setValue(value)
        }
        onChanged()
    }

    fun setGroupValue(value: String?) {
        if (value == null) {
            groupValue.setValue(listOf(""))
        } else {
            groupValue.setValue(listOf(value))
        }
        onChanged()
    }

    fun setGroupValues(value: List<String>) {
        groupValue.setValue(value)
        onChanged()
    }

    fun setTypeValue(value: Int) {
        typeValue.setValue(value)
        onChanged()
    }

    fun setStatusValue(value: Int) {
        statusValue.setValue(value)
        onChanged()
    }

    fun setRangeValue(value: Int) {
        when (value) {
            0 -> this.rangeValue.setValue(DateFilter.DateRange.ALL)
            1 -> this.rangeValue.setValue(DateFilter.DateRange.PERMANENT)
            2 -> this.rangeValue.setValue(DateFilter.DateRange.TODAY)
            3 -> this.rangeValue.setValue(DateFilter.DateRange.TOMORROW)
        }
        onChanged()
    }

    fun getOriginal(): List<Reminder> {
        return original
    }

    fun setOriginal(original: MutableList<Reminder>) {
        this.original = original
        onChanged()
    }

    private fun onChanged() {
        val list = ArrayList<Reminder>()
        for (reminder in original) {
            if (mFilter != null) {
                if (mFilter!!.filter(reminder)) list.add(reminder)
            } else
                list.add(reminder)
        }
        mCallback?.onChanged(list)
    }

    fun remove(reminder: Reminder) {
        if (original.contains(reminder)) {
            original.remove(reminder)
            onChanged()
        }
    }
}
